src/access.rs

```diff
@@ -0,0 +1,361 @@
+		let watched_outputs = Mutex::new(Vec::new());
+		let queued_outputs = Mutex::new(Vec::new());
+		let last_sync_height = tokio::sync::Mutex::new(None);
+		Self {
+			blockchain,
+			wallet,
+			queued_transactions,
+			watched_transactions,
+			queued_outputs,
+			watched_outputs,
+			last_sync_height,
+			logger,
+		}
+	}
+
+	pub(crate) async fn sync_wallet(&self) -> Result<(), Error> {
+		let sync_options = SyncOptions { progress: None };
+
+		self.wallet
+			.lock()
+			.unwrap()
+			.sync(&self.blockchain, sync_options)
+			.map_err(|e| Error::Bdk(e))?;
+
+		Ok(())
+	}
+
+	pub(crate) async fn sync(&self, confirmables: Vec<&(dyn Confirm + Sync)>) -> Result<(), Error> {
+		let client = &*self.blockchain;
+
+		let cur_height = client.get_height().await?;
+
+		let mut locked_last_sync_height = self.last_sync_height.lock().await;
+		if cur_height >= locked_last_sync_height.unwrap_or(0) {
+			self.sync_best_block_updated(&confirmables, cur_height, &mut locked_last_sync_height)
+				.await?;
+			self.sync_transactions_confirmed(&confirmables).await?;
+			self.sync_transaction_unconfirmed(&confirmables).await?;
+		}
+		// TODO: check whether new outputs have been registered by now and process them
+		Ok(())
+	}
+
+	async fn sync_best_block_updated(
+		&self, confirmables: &Vec<&(dyn Confirm + Sync)>, cur_height: u32,
+		locked_last_sync_height: &mut tokio::sync::MutexGuard<'_, Option<u32>>,
+	) -> Result<(), Error> {
+		let client = &*self.blockchain;
+
+		// Inform the interface of the new block.
+		let cur_block_header = client.get_header(cur_height).await?;
+		for c in confirmables {
+			c.best_block_updated(&cur_block_header, cur_height);
+		}
+
+		**locked_last_sync_height = Some(cur_height);
+		Ok(())
+	}
+
+	async fn sync_transactions_confirmed(
+		&self, confirmables: &Vec<&(dyn Confirm + Sync)>,
+	) -> Result<(), Error> {
+		let client = &*self.blockchain;
+
+		// First, check the confirmation status of registered transactions as well as the
+		// status of dependent transactions of registered outputs.
+
+		let mut confirmed_txs = Vec::new();
+
+		// Check in the current queue, as well as in registered transactions leftover from
+		// previous iterations.
+		let registered_txs: HashSet<Txid> = {
+			let locked_queued_transactions = self.queued_transactions.lock().unwrap();
+			let locked_watched_transactions = self.watched_transactions.lock().unwrap();
+			locked_watched_transactions
+				.iter()
+				.chain(locked_queued_transactions.iter())
+				.cloned()
+				.collect()
+		};
+
+		// Remember all registered but unconfirmed transactions for future processing.
+		let mut unconfirmed_registered_txs = Vec::new();
+
+		for txid in registered_txs {
+			if let Some(tx_status) = client.get_tx_status(&txid).await? {
+				if tx_status.confirmed {
+					if let Some(tx) = client.get_tx(&txid).await? {
+						if let Some(block_height) = tx_status.block_height {
+							let block_header = client.get_header(block_height).await?;
```

Comment:
Yes it is, looking at these two lines only. However, if the TX would have been reorged-out the next call to `get_merkle_proof` would fail. I now added a explicit check that the height didn't change inbetween.

On top of that we could try to query the block by hash here, just opened an upstream PR that would give us that option (https://github.com/bitcoindevkit/rust-esplora-client/pull/17).

Reply:
> Right, yea, we'll need upstream to fix this. A simpler fix, if its available, would be to extract the block header from the merkle proof. 

Unfortunately, it is not available. While there is an alternate (so far unimplemented) API method that would return the merkle proof in Bitcoind's `merkleblock` format which would feature the header, this then in turn wouldn't give us the `pos`, IIUC.